bigtable/client/row_range.cc

```diff
@@ -0,0 +1,446 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/row_range.h"
+
+#include <iostream>
+#include <thread>
+
+#include <absl/memory/memory.h>
+#include <absl/strings/str_cat.h>
+
+namespace btproto = ::google::bigtable::v2;
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+const std::string RowRange::SAMPLE_KEY = "";
```

Comment:
I am not sure why this is a "SAMPLE".  Is it that the empty string is an illegal row key and therefore we can use it as an special value here?
If so, the more appropriate name would be "INVALID_ROW_KEY".

Reply:
Empty key for 'end' has special significance: it implies infinity (+infinity). Now, to keep things simpler, we infer the 'infinity' by looking only at the end_key_case. This makes code simpler, as we don't have to look at the key value---this look up would be conditioned on end_key_case. So, now our assumption is: NOT_SET for end_key means (+ve) infinity, and the  right side in such a case of infinity is OPEN. 

I was thinking of using the same representation for start_key, i.e., not allowing the user to set start to "". In that case,  NOT_SET would  means the interval is  CLOSED at the left. This would unify the representation and remove the confusion, at the cost of checks for start values. When user specified open_interval("", x), the interval would be right-open  ["", x). 

Now, regarding SAMPLE_KEY="", this sample key could be any value x.  I used "" just to have a printed range [..) which would be a better representation for empty range than  [FOO..FOO).